### ISSUE TITLE
Inconsistency of import cycles behind star exports

### ISSUE THREAD

**Issue by bojavou** [tc39/ecma262#3129]
**Description:**

There is some inconsistency where an import cycle exists behind a star export.

```js
// entry.mjs
import { value } from './aggregate.mjs'

// aggregate.mjs
export * from './cycle.mjs'
export * from './binding.mjs'

// cycle.mjs
export { value } from './entry.mjs'

// binding.mjs
export const value = 'binding'
```

V8 calls this an undefined export:

>SyntaxError: The requested module './entry.mjs' does not provide an export named 'value'

SpiderMonkey calls it ambiguous:

>SyntaxError: ambiguous import: value

By my reading of the spec, this should actually succeed. The cycle should be overridden by the import path that resolves successfully.

I wonder if something should be done to align things. I don't know if that means a spec change or implementation changes.

**Explanation:**

The relevant logic is in [`ResolveExport`](https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#sec-resolveexport).

Line 3 detects cycles. If a cycle is detected, the method returns `null`.

Down in lines 8-9, star exports are handled. First the result variable `starResolution` is initialized to `null`:

```
8. Let starResolution be null.
```

Then there's a recursive call to `ResolveExport`:

```
9.b. Let resolution be importedModule.ResolveExport(exportName, resolveSet).
```

If the star export finds a cycle, `resolution` will now be `null`.

The bulk of the logic only happens if `resolution` is not `null`.

```
9.d. If resolution is not null, then
```

Under this branch is the only place the result variable is changed.

```
9.d.ii. If starResolution is null, then
    1. Set starResolution to resolution.
```

So this situation should return a successful binding:

1. The first star export finds a cycle, which returns `null`, and therefore leaves `starResolution` as `null`.
1. The second star export successfully resolves to a binding. Since `starResolution` is still `null`, the binding is stored as the tentative result.
1. If there are no other star exports that create ambiguity, the resolved binding is the result.

**Demonstration:**

[`star-export-cycle`](https://github.com/bojavou/star-export-cycle) has a prepared demonstration. Run `entry.mjs` to reproduce.

It can be run in the browser here. The error will show up in the console:

https://bojavou.github.io/star-export-cycle/

**eshost Output:**

These are all the engines I was able to get running.

```
$ eshost entry.mjs
#### engine262

SyntaxError: Could not resolve import 'value' from /tmp/MVHct9RYFwmrY3ei7i4F/cycle.mjs

#### JavaScriptCore

SyntaxError: Indirectly exported binding name 'value' is not found.

#### SpiderMonkey

SyntaxError: ambiguous import: value

#### V8

SyntaxError: The requested module './entry.mjs' does not provide an export named 'value'

#### XS

SyntaxError: (host): import value not found
```

**Comment by bojavou** [comment]
(eshost is so nice!)

**Comment by bojavou** [comment]
Had my demonstration wrong. Just updated everything.

**Comment by nicolo-ribaudo** [comment]
You are correct that it does not throw when running `aggregate.ResolveExport("value")`. However, it throws when executing `cycle.InitializeEnvironment()`, at step 1.b.
